log the applied patches for each package if CONFIG_BUILD_LOG is set

diff --git a/include/quilt.mk b/include/quilt.mk
--- a/include/quilt.mk
+++ b/include/quilt.mk
@@ -136,6 +136,13 @@ define Quilt/Template
 				$(QUILT_CMD) top >/dev/null 2>&1; \
 			fi \
 		); \
+		if [ "$(BUILD_LOG)" = "1" ]; then \
+			log_dir=$(BUILD_LOG_DIR)/patches; \
+			package=$$$$(basename $(1)); \
+			mkdir -p $$$$log_dir; \
+			log=$$$$log_dir/$$$$package.txt; \
+			cat $(1)/patches/series > $$$$log; \
+		fi \
 	fi
 	touch "$$@"
 
diff --git a/package/Makefile b/package/Makefile
--- a/package/Makefile
+++ b/package/Makefile
@@ -57,6 +57,17 @@ ifdef CONFIG_USE_MKLIBS
   endef
 endif
 
+ifeq ($(BUILD_LOG),1)
+  define install_patches_info
+	>$(BUILD_LOG_DIR)/applied_patches
+	for file in `find $(BUILD_LOG_DIR)/patches -name '*.txt'`; do \
+		name=`basename $$file | sed -e 's/\.txt$$//'`; \
+		echo "$$name:" >>$(BUILD_LOG_DIR)/applied_patches; \
+		cat $$file >>$(BUILD_LOG_DIR)/applied_patches; \
+	done
+  endef
+endif
+
 # where to build (and put) .ipk packages
 OPKG:= \
   IPKG_NO_SCRIPT=1 \
@@ -120,6 +131,7 @@ PACKAGE_INSTALL_FILES:= \
 	rm -f $(TARGET_DIR)/usr/lib/opkg/info/*.prerm*
 	$(if $(CONFIG_CLEAN_IPKG),rm -rf $(TARGET_DIR)/usr/lib/opkg)
 	$(call mklibs)
+	$(call install_patches_info)
 
 PASSOPT=""
 PASSARG=""
diff --git a/scripts/patch-kernel.sh b/scripts/patch-kernel.sh
--- a/scripts/patch-kernel.sh
+++ b/scripts/patch-kernel.sh
@@ -19,6 +19,23 @@ if [ ! -d "${patchdir}" ] ; then
     exit 1
 fi
     
+BUILD_LOG=0
+if [ ! -z $TOPDIR ]; then
+	if cat $TOPDIR/.config | grep "CONFIG_BUILD_LOG=y" &>/dev/null; then
+		BUILD_LOG=1
+	fi
+
+	if [ $BUILD_LOG -eq 1 ]; then
+		log_dir=$TOPDIR/logs/patches/
+		package=$(basename $PWD)
+		mkdir -p $log_dir
+
+		log=$log_dir/$package.txt
+
+		>$log
+	fi
+fi
+
 for i in ${patchdir}/${patchpattern} ; do 
     case "$i" in
 	*.gz)
@@ -42,6 +59,8 @@ for i in ${patchdir}/${patchpattern} ; d
         echo "Patch failed!  Please fix $i!"
 	exit 1
     fi
+
+    [ $BUILD_LOG -eq 1 ] && { echo $(basename $i) >> $log; }
 done
 
 # Check for rejects...
